nav2_system_tests/src/costmap_filters/costmap_filters_tests_launch.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2020 Samsung Research Russia
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
```

Comment:
This is not a system test. There's no full navigation test here to show that the robot wont travel in the keepout zone. This is just an integration test, please move to costmap_2d. 

How does this test tangibly differ from the costmap_2d tests? How do you test rolling / local costmaps?

Reply:
> please move to costmap_2d.

Unfortunately, the test can't be moved into nav2_costmap_2d integration tests because it is using NavFN planner. NavFN planner itself depends on `nav2_costmap_2d` package. If add NavFN planner to costamp's package dependencies this will lead to error message because of cycle-dependency, like:
```
[0.411s] ERROR:colcon:colcon build: Unable to order packages topologically
```

> There's no full navigation test here to show that the robot wont travel in the keepout zone.

The testcase is designed to step-by-step move robot on produced by NavFN plan and check on each step that robot won't enter a keepout area. Also it is verifying that robot can reach to valid goal _bypassing_ all keepout zones and won't move to the goal set inside a keepout zone. This test is intended to be not a real-time test, it is turn-based iterative test (similar to `nav2_system_tests/src/planning` does).

> How do you test rolling / local costmaps?

During this test, Costmap2DROS has no sensors data, so it restricts the window area `(min_i, min_j, max_i, max_j)` to robot's footprint. On each testing step, Costmap2DROS calling `keepout_filter` plugin with changing window coordinates. So, it is also covers a rolling window.

> How does this test tangibly differ from the costmap_2d tests? 

It differs from keepout unit test because of unit test does not check robot moving with path planner, it just verifies built costmaps in static. In opposite, `nav2_system_tests/src/costmap_filters/test_keepout_filter.cpp` checks all keepout filter functionality from outside and acts as a system test in iterative world having Costmap2DROS, Path Planner and Map Server onboard. Why do not retain this test?

I also would like agree with you, that we need to have full navigation stack based real-time test for conditions closer to real world. I will try to make it in the update. But I still consider, that iterative `nav2_system_tests/src/costmap_filters/test_keepout_filter` test checks all keepout filter functionality from outside, but in a slightly non-standard way.

Reply:
> he test can't be moved into nav2_costmap_2d integration tests because it is using NavFN planner.